File: docs/breez-sdk/snippets/rust/src/send_payment.rs
use anyhow::Result;
use breez_sdk_spark::*;
use log::info;

async fn prepare_send_payment_lightning_bolt11(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: prepare-send-payment-lightning-bolt11
    let payment_request = "<bolt11 invoice>".to_string();
    // Optionally set the amount you wish to pay the receiver
    let optional_amount_sats = Some(5_000);

    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request,
            amount: optional_amount_sats,
            token_identifier: None,
            conversion_options: None,
            fee_policy: None,
        })
        .await?;

    // If the fees are acceptable, continue to create the Send Payment
    if let SendPaymentMethod::Bolt11Invoice {
        spark_transfer_fee_sats,
        lightning_fee_sats,
        ..
    } = prepare_response.payment_method
    {
        // Fees to pay via Lightning
        info!("Lightning Fees: {lightning_fee_sats} sats");
        // Or fees to pay (if available) via a Spark transfer
        info!("Spark Transfer Fees: {spark_transfer_fee_sats:?} sats");
    }
    // ANCHOR_END: prepare-send-payment-lightning-bolt11
    Ok(())
}

async fn prepare_send_payment_onchain(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: prepare-send-payment-onchain
    let payment_request = "<bitcoin address>".to_string();
    // Set the amount you wish to pay the receiver
    let amount_sats = Some(50_000);

    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request,
            amount: amount_sats,
            token_identifier: None,
            conversion_options: None,
            fee_policy: None,
        })
        .await?;

    // Review the fee quote for each confirmation speed
    if let SendPaymentMethod::BitcoinAddress { fee_quote, .. } = &prepare_response.payment_method {
        info!("Slow fee: {} sats", fee_quote.speed_slow.total_fee_sat());
        info!("Medium fee: {} sats", fee_quote.speed_medium.total_fee_sat());
        info!("Fast fee: {} sats", fee_quote.speed_fast.total_fee_sat());
    }
    // ANCHOR_END: prepare-send-payment-onchain
    Ok(())
}

async fn prepare_send_payment_spark_address(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: prepare-send-payment-spark-address
    let payment_request = "<spark address>".to_string();
    // Set the amount you wish to pay the receiver
    let amount_sats = Some(50_000);

    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request,
            amount: amount_sats,
            token_identifier: None,
            conversion_options: None,
            fee_policy: None,
        })
        .await?;

    // If the fees are acceptable, continue to create the Send Payment
    if let SendPaymentMethod::SparkAddress { fee, .. } = prepare_response.payment_method {
        info!("Fees: {} sats", fee);
    }
    // ANCHOR_END: prepare-send-payment-spark-address
    Ok(())
}

async fn prepare_send_payment_spark_invoice(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: prepare-send-payment-spark-invoice
    let payment_request = "<spark invoice>".to_string();
    // Optionally set the amount you wish to pay the receiver
    let optional_amount_sats = Some(50_000);

    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request,
            amount: optional_amount_sats,
            token_identifier: None,
            conversion_options: None,
            fee_policy: None,
        })
        .await?;

    // If the fees are acceptable, continue to create the Send Payment
    if let SendPaymentMethod::SparkInvoice { fee, .. } = prepare_response.payment_method {
        info!("Fees: {} sats", fee);
    }
    // ANCHOR_END: prepare-send-payment-spark-invoice
    Ok(())
}

async fn prepare_send_payment_token_conversion(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: prepare-send-payment-with-conversion
    let payment_request = "<payment request>".to_string();
    // Set to use token funds to pay via conversion
    let optional_max_slippage_bps = Some(50);
    let optional_completion_timeout_secs = Some(30);
    let conversion_options = Some(ConversionOptions {
        conversion_type: ConversionType::ToBitcoin {
            from_token_identifier: "<token identifier>".to_string(),
        },
        max_slippage_bps: optional_max_slippage_bps,
        completion_timeout_secs: optional_completion_timeout_secs,
    });

    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request,
            amount: None,
            token_identifier: None,
            conversion_options,
            fee_policy: None,
        })
        .await?;

    // If the fees are acceptable, continue to create the Send Payment
    if let Some(conversion_estimate) = &prepare_response.conversion_estimate {
        info!("Estimated conversion amount: {} token base units", conversion_estimate.amount);
        info!("Estimated conversion fee: {} token base units", conversion_estimate.fee);
    }
    // ANCHOR_END: prepare-send-payment-with-conversion
    Ok(())
}

async fn send_payment_lightning_bolt11(
    sdk: &BreezSdk,
    prepare_response: PrepareSendPaymentResponse,
) -> Result<()> {
    // ANCHOR: send-payment-lightning-bolt11
    let options = Some(SendPaymentOptions::Bolt11Invoice {
        prefer_spark: false,
        completion_timeout_secs: Some(10),
    });
    let optional_idempotency_key = Some("<idempotency key uuid>".to_string());
    let send_response = sdk
        .send_payment(SendPaymentRequest {
            prepare_response,
            options,
            idempotency_key: optional_idempotency_key,
        })
        .await?;
    let payment = send_response.payment;
    info!("Payment: {payment:?}");
    // ANCHOR_END: send-payment-lightning-bolt11
    Ok(())
}

async fn send_payment_onchain(
    sdk: &BreezSdk,
    prepare_response: PrepareSendPaymentResponse,
) -> Result<()> {
    // ANCHOR: send-payment-onchain
    // Select the confirmation speed for the on-chain transaction
    let options = Some(SendPaymentOptions::BitcoinAddress {
        confirmation_speed: OnchainConfirmationSpeed::Medium,
    });
    let optional_idempotency_key = Some("<idempotency key uuid>".to_string());
    let send_response = sdk
        .send_payment(SendPaymentRequest {
            prepare_response,
            options,
            idempotency_key: optional_idempotency_key,
        })
        .await?;
    let payment = send_response.payment;
    info!("Payment: {payment:?}");
    // ANCHOR_END: send-payment-onchain
    Ok(())
}

async fn send_payment_spark(
    sdk: &BreezSdk,
    prepare_response: PrepareSendPaymentResponse,
) -> Result<()> {
    // ANCHOR: send-payment-spark
    let optional_idempotency_key = Some("<idempotency key uuid>".to_string());
    let send_response = sdk
        .send_payment(SendPaymentRequest {
            prepare_response,
            options: None,
            idempotency_key: optional_idempotency_key,
        })
        .await?;
    let payment = send_response.payment;
    info!("Payment: {payment:?}");
    // ANCHOR_END: send-payment-spark
    Ok(())
}

async fn prepare_send_payment_fees_included(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: prepare-send-payment-fees-included
    // By default (FeePolicy::FeesExcluded), fees are added on top of the amount.
    // Use FeePolicy::FeesIncluded to deduct fees from the amount instead.
    // The receiver gets amount minus fees.
    let payment_request = "<payment request>".to_string();
    let amount_sats = Some(50_000);

    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request,
            amount: amount_sats,
            token_identifier: None,
            conversion_options: None,
            fee_policy: Some(FeePolicy::FeesIncluded),
        })
        .await?;

    // The response shows the fee policy used
    info!("Fee policy: {:?}", prepare_response.fee_policy);
    info!("Amount: {}", prepare_response.amount);
    // The receiver gets amount - fees (fees are available in prepare_response.payment_method)
    // ANCHOR_END: prepare-send-payment-fees-included
    Ok(())
}

File: docs/breez-sdk/snippets/csharp/SendPayment.cs
using Breez.Sdk.Spark;

namespace BreezSdkSnippets
{
    class SendPayment
    {
        async Task PrepareSendPaymentLightningBolt11(BreezSdk sdk)
        {
            // ANCHOR: prepare-send-payment-lightning-bolt11
            var paymentRequest = "<bolt11 invoice>";
            // Optionally set the amount you wish to pay the receiver
            ulong? optionalAmountSats = 5_000UL;

            var request = new PrepareSendPaymentRequest(
                paymentRequest: paymentRequest,
                amount: optionalAmountSats,
                tokenIdentifier: null,
                conversionOptions: null,
                feePolicy: null
            );
            var prepareResponse = await sdk.PrepareSendPayment(request: request);

            // If the fees are acceptable, continue to create the Send Payment
            if (prepareResponse.paymentMethod is SendPaymentMethod.Bolt11Invoice bolt11Method)
            {
                // Fees to pay via Lightning
                var lightningFeeSats = bolt11Method.lightningFeeSats;
                // Or fees to pay (if available) via a Spark transfer
                var sparkTransferFeeSats = bolt11Method.sparkTransferFeeSats;
                Console.WriteLine($"Lightning Fees: {lightningFeeSats} sats");
                Console.WriteLine($"Spark Transfer Fees: {sparkTransferFeeSats} sats");
            }
            // ANCHOR_END: prepare-send-payment-lightning-bolt11
        }

        async Task PrepareSendPaymentOnchain(BreezSdk sdk)
        {
            // ANCHOR: prepare-send-payment-onchain
            var paymentRequest = "<bitcoin address>";
            // Set the amount you wish to pay the receiver
            ulong? amountSats = 50_000UL;

            var request = new PrepareSendPaymentRequest(
                paymentRequest: paymentRequest,
                amount: amountSats,
                tokenIdentifier: null,
                conversionOptions: null,
                feePolicy: null
            );
            var prepareResponse = await sdk.PrepareSendPayment(request: request);

            // Review the fee quote for each confirmation speed
            if (prepareResponse.paymentMethod is SendPaymentMethod.BitcoinAddress bitcoinMethod)
            {
                var feeQuote = bitcoinMethod.feeQuote;
                var slowFeeSats = feeQuote.speedSlow.userFeeSat + feeQuote.speedSlow.l1BroadcastFeeSat;
                var mediumFeeSats = feeQuote.speedMedium.userFeeSat + feeQuote.speedMedium.l1BroadcastFeeSat;
                var fastFeeSats = feeQuote.speedFast.userFeeSat + feeQuote.speedFast.l1BroadcastFeeSat;
                Console.WriteLine($"Slow fee: {slowFeeSats} sats");
                Console.WriteLine($"Medium fee: {mediumFeeSats} sats");
                Console.WriteLine($"Fast fee: {fastFeeSats} sats");
            }
            // ANCHOR_END: prepare-send-payment-onchain
        }

        async Task PrepareSendPaymentSparkAddress(BreezSdk sdk)
        {
            // ANCHOR: prepare-send-payment-spark-address
            var paymentRequest = "<spark address>";
            // Set the amount you wish to pay the receiver
            ulong? amountSats = 50_000UL;

            var request = new PrepareSendPaymentRequest(
                paymentRequest: paymentRequest,
                amount: amountSats,
                tokenIdentifier: null,
                conversionOptions: null,
                feePolicy: null
            );
            var prepareResponse = await sdk.PrepareSendPayment(request: request);

            // If the fees are acceptable, continue to create the Send Payment
            if (prepareResponse.paymentMethod is SendPaymentMethod.SparkAddress sparkMethod)
            {
                var fee = sparkMethod.fee;
                Console.WriteLine($"Fees: {fee} sats");
            }
            // ANCHOR_END: prepare-send-payment-spark-address
        }

        async Task PrepareSendPaymentSparkInvoice(BreezSdk sdk)
        {
            // ANCHOR: prepare-send-payment-spark-invoice
            var paymentRequest = "<spark invoice>";
            // Optionally set the amount you wish to pay the receiver
            ulong? optionalAmountSats = 50_000UL;

            var request = new PrepareSendPaymentRequest(
                paymentRequest: paymentRequest,
                amount: optionalAmountSats,
                tokenIdentifier: null,
                conversionOptions: null,
                feePolicy: null
            );
            var prepareResponse = await sdk.PrepareSendPayment(request: request);

            // If the fees are acceptable, continue to create the Send Payment
            if (prepareResponse.paymentMethod is SendPaymentMethod.SparkInvoice sparkInvoiceMethod)
            {
                var fee = sparkInvoiceMethod.fee;
                Console.WriteLine($"Fees: {fee} sats");
            }
            // ANCHOR_END: prepare-send-payment-spark-invoice
        }

        async Task PrepareSendPaymentFeesIncluded(BreezSdk sdk)
        {
            // ANCHOR: prepare-send-payment-fees-included
            // By default (FeePolicy.FeesExcluded), fees are added on top of the amount.
            // Use FeePolicy.FeesIncluded to deduct fees from the amount instead.
            // The receiver gets amount minus fees.
            var paymentRequest = "<payment request>";
            ulong? amountSats = 50_000UL;

            var request = new PrepareSendPaymentRequest(
                paymentRequest: paymentRequest,
                amount: amountSats,
                tokenIdentifier: null,
                conversionOptions: null,
                feePolicy: FeePolicy.FeesIncluded
            );
            var prepareResponse = await sdk.PrepareSendPayment(request: request);

            // The response shows the fee policy used
            Console.WriteLine($"Fee policy: {prepareResponse.feePolicy}");
            Console.WriteLine($"Amount: {prepareResponse.amount}");
            // The receiver gets amount - fees (fees are available in prepareResponse.paymentMethod)
            // ANCHOR_END: prepare-send-payment-fees-included
        }

        async Task PrepareSendPaymentTokenConversion(BreezSdk sdk)
        {
            // ANCHOR: prepare-send-payment-with-conversion
            var paymentRequest = "<payment request>";
            // Set to use token funds to pay via conversion
            var optionalMaxSlippageBps = 50U;
            var optionalCompletionTimeoutSecs = 30U;
            var conversionOptions = new ConversionOptions(
                conversionType: new ConversionType.ToBitcoin(
                    fromTokenIdentifier: "<token identifier>"
                ),
                maxSlippageBps: optionalMaxSlippageBps,
                completionTimeoutSecs: optionalCompletionTimeoutSecs
            );

            var request = new PrepareSendPaymentRequest(
                paymentRequest: paymentRequest,
                amount: null,
                tokenIdentifier: null,
                conversionOptions: conversionOptions,
                feePolicy: null
            );
            var prepareResponse = await sdk.PrepareSendPayment(request: request);

            // If the fees are acceptable, continue to create the Send Payment
            if (prepareResponse.conversionEstimate != null)
            {
                Console.WriteLine("Estimated conversion amount: " +
                    $"{prepareResponse.conversionEstimate.amount} token base units");
                Console.WriteLine("Estimated conversion fee: " +
                    $"{prepareResponse.conversionEstimate.fee} token base units");
            }
            // ANCHOR_END: prepare-send-payment-with-conversion
        }

        async Task SendPaymentLightningBolt11(BreezSdk sdk, PrepareSendPaymentResponse prepareResponse)
        {
            // ANCHOR: send-payment-lightning-bolt11
            var options = new SendPaymentOptions.Bolt11Invoice(
                preferSpark: false,
                completionTimeoutSecs: 10
            );
            var optionalIdempotencyKey = "<idempotency key uuid>";
            var request = new SendPaymentRequest(
                prepareResponse: prepareResponse,
                options: options,
                idempotencyKey: optionalIdempotencyKey
            );
            var sendResponse = await sdk.SendPayment(request: request);
            var payment = sendResponse.payment;
            // ANCHOR_END: send-payment-lightning-bolt11
        }

        async Task SendPaymentOnchain(BreezSdk sdk, PrepareSendPaymentResponse prepareResponse)
        {
            // ANCHOR: send-payment-onchain
            // Select the confirmation speed for the on-chain transaction
            var options = new SendPaymentOptions.BitcoinAddress(
                confirmationSpeed: OnchainConfirmationSpeed.Medium
            );
            var optionalIdempotencyKey = "<idempotency key uuid>";
            var request = new SendPaymentRequest(
                prepareResponse: prepareResponse,
                options: options,
                idempotencyKey: optionalIdempotencyKey
            );
            var sendResponse = await sdk.SendPayment(request: request);
            var payment = sendResponse.payment;
            // ANCHOR_END: send-payment-onchain
        }

        async Task SendPaymentSpark(BreezSdk sdk, PrepareSendPaymentResponse prepareResponse)
        {
            // ANCHOR: send-payment-spark
            var optionalIdempotencyKey = "<idempotency key uuid>";
            var request = new SendPaymentRequest(
                prepareResponse: prepareResponse,
                idempotencyKey: optionalIdempotencyKey
            );
            var sendResponse = await sdk.SendPayment(request: request);
            var payment = sendResponse.payment;
            // ANCHOR_END: send-payment-spark
        }
    }
}
